openmeter/productcatalog/plan/service/plan.go

```diff
@@ -0,0 +1,295 @@
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/samber/lo"
+
+	"github.com/openmeterio/openmeter/openmeter/productcatalog/plan"
+	"github.com/openmeterio/openmeter/pkg/models"
+	"github.com/openmeterio/openmeter/pkg/pagination"
+)
+
+func (s service) ListPlans(ctx context.Context, params plan.ListPlansInput) (pagination.PagedResponse[plan.Plan], error) {
+	if err := params.Validate(); err != nil {
+		return pagination.PagedResponse[plan.Plan]{}, fmt.Errorf("invalid list Plans params: %w", err)
+	}
+
+	return s.adapter.ListPlans(ctx, params)
+}
+
+func (s service) CreatePlan(ctx context.Context, params plan.CreatePlanInput) (*plan.Plan, error) {
+	if err := params.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid create Plan params: %w", err)
+	}
+
+	logger := s.logger.With(
+		"operation", "create",
+		"namespace", params.Namespace,
+		"plan.key", params.Key,
+	)
+
+	logger.Debug("creating Plan")
+
+	p, err := s.adapter.CreatePlan(ctx, params)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Plan: %w", err)
+	}
+
+	logger.With("plan.id", p.ID).Debug("Plan created")
+
+	return p, nil
+}
+
+func (s service) DeletePlan(ctx context.Context, params plan.DeletePlanInput) error {
+	if err := params.Validate(); err != nil {
+		return fmt.Errorf("invalid delete Plan params: %w", err)
+	}
+
+	logger := s.logger.With(
+		"operation", "delete",
+		"namespace", params.Namespace,
+		"plan.id", params.ID,
+	)
+
+	// TODO(chrisgacsal): add check which makes sure that Plans with active Subscriptions are not deleted.
+
+	logger.Debug("deleting Plan")
+
+	err := s.adapter.DeletePlan(ctx, params)
+	if err != nil {
+		return fmt.Errorf("failed to delete Plan: %w", err)
+	}
+
+	logger.Debug("Plan deleted")
```

Comment:
just a thought, but wouldn't traces be better suited for this kind of debugging?

Reply:
Agreed. Thats my plan to do. Debug logging is there until the instrumentation is added.